citadel_workspace_types/src/lib.rs

```diff
@@ -154,6 +154,257 @@ pub struct PeerRegisterFailure {
     pub request_id: Option<Uuid>,
 }
 
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub struct GroupChannelCreateSuccess {
+    pub cid: u64,
+    pub group_key: MessageGroupKey,
+    pub request_id: Option<Uuid>,
+}
+
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub struct GroupChannelCreateFailure {
+    pub cid: u64,
+    pub group_key: MessageGroupKey,
+    pub message: String,
+    pub request_id: Option<Uuid>,
+}
+
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub struct GroupBroadcastHandleFailure {
+    pub cid: u64,
+    pub message: String,
+    pub request_id: Option<Uuid>,
+}
+
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub struct GroupCreateSuccess {
+    pub cid: u64,
+    pub group_key: MessageGroupKey,
+    pub request_id: Option<Uuid>,
+}
+
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub struct GroupCreateFailure {
+    pub cid: u64,
+    pub message: String,
+    pub request_id: Option<Uuid>,
+}
+
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub struct GroupLeaveSuccess {
+    pub cid: u64,
+    pub group_key: MessageGroupKey,
+    pub request_id: Option<Uuid>,
+}
+
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub struct GroupLeaveFailure {
+    pub cid: u64,
+    pub message: String,
+    pub request_id: Option<Uuid>,
+}
+
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub struct GroupEndSuccess {
+    pub cid: u64,
+    pub group_key: MessageGroupKey,
+    pub request_id: Option<Uuid>,
+}
+
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub struct GroupEndFailure {
+    pub cid: u64,
+    pub message: String,
+    pub request_id: Option<Uuid>,
+}
+
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub struct GroupEnded {
+    pub cid: u64,
+    pub group_key: MessageGroupKey,
+    pub success: bool,
+    pub request_id: Option<Uuid>,
+}
+
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub struct GroupLeft {
+    pub cid: u64,
+    pub group_key: MessageGroupKey,
+    pub success: bool,
+    pub message: String,
+    pub request_id: Option<Uuid>,
+}
+
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub struct GroupMessageReceived {
+    pub cid: u64,
+    pub peer_cid: u64,
+    pub message: Vec<u8>,
```

Comment:
Consider using a `SecBuffer` instead that way the earlier `buffer.into_buffer()` call doesn't need an additional `.into()` call that appears to be cloning under the surface. Zero-copy is the way to go when forwarding messages between places in memory.